### ISSUE TITLE
Incorrect loss values with SVMOutput/Hinge loss

### ISSUE THREAD

**Issue by jmerkow** [apache/incubator-mxnet#6466]
I am testing hinge loss (SVMOutput) and it appears to be producing negative loss values which should not be possible given it has the form `max(0, --)`. Is MXNet using an alternative calculation for this loss function that allows negative values?  Perhaps this max occurs after `eval_metric` updates?

Here is some code, references [gist](https://gist.github.com/jmerkow/1669eed6c7d74bee13db402df495f45f):

```
from gist.googlenet import get_net
from gist.PR_loggers import LogMetricsCallback, TrainingLogger


batch_size = 64
base_lr = 0.0001
kv = mx.kvstore.create('local')

train_data = ...
eval_data = ...
# whatever data you like, for example:
# train_data = SyntheticDataIter(2, [batch_size,1,224,224], 15000, 'float')
# eval_data = SyntheticDataIter(2, [1,1,224,224], 250, 'float')

batch_end_callbacks = [TrainingLogger(frequent=50, batch_size=batch_size, prefix='Train-')]
eval_end_callbacks = [LogMetricsCallback(prefix='Validation-')]
initializer = mx.init.Xavier()

optimizer_params = {
        'learning_rate': base_lr,
        'momentum' : .5,
        'wd' : 0}
net = mx.symbol.SVMOutput(data=get_net(2), name='softmax')
mod = mx.mod.Module(symbol=net, context=[mx.gpu(0)])
mod.fit(
        train_data,
        eval_data,
        
        eval_metric=['loss', 'accuracy'],
        kvstore = kv,
        
        optimizer='sgd',
        optimizer_params=optimizer_params,
        
        batch_end_callback=batch_end_callbacks,
        eval_end_callback=eval_end_callbacks,
        
        initializer=initializer,
        num_epoch=5)
```

This gives output which includes negative loss values which should not be possible.

> 2017-05-26 13:41:34,380 Node[0] Epoch[0] Batch[50] Train-lr=0.0001 Train-iter_per_sec=1.08 Train-samples_per_sec=69.21 Train-loss=0.01018502615 Train-accuracy=0.4840686275
> 2017-05-26 13:42:20,844 Node[0] Epoch[0] Batch[100] Train-lr=0.0001 Train-iter_per_sec=1.08 Train-samples_per_sec=68.87 Train-loss=-0.0001527548686 Train-accuracy=0.545625
> 2017-05-26 13:43:07,365 Node[0] Epoch[0] Batch[150] Train-lr=0.0001 Train-iter_per_sec=1.07 Train-samples_per_sec=68.79 Train-loss=0.0001172707253 Train-accuracy=0.5746875
> 2017-05-26 13:43:53,881 Node[0] Epoch[0] Batch[200] Train-lr=0.0001 Train-iter_per_sec=1.07 Train-samples_per_sec=68.8 Train-loss=0.0001611461397 Train-accuracy=0.5728125
> 2017-05-26 13:44:26,449 Node[0] Epoch[0] Train-loss=-0.000160
> 2017-05-26 13:44:26,451 Node[0] Epoch[0] Train-accuracy=0.599107
> 2017-05-26 13:44:26,451 Node[0] Epoch[0] Time-cost=219.891
> 2017-05-26 13:44:28,393 Node[0] Epoch[0] Validation-loss=-0.02287905058 Validation-accuracy=0.51171875
> 2017-05-26 13:45:15,985 Node[0] Epoch[1] Batch[50] Train-lr=0.0001 Train-iter_per_sec=1.85 Train-samples_per_sec=68.76 Train-loss=-0.0005699406987 Train-accuracy=0.6253063725
> 2017-05-26 13:46:02,525 Node[0] Epoch[1] Batch[100] Train-lr=0.0001 Train-iter_per_sec=1.07 Train-samples_per_sec=68.76 Train-loss=-0.0001366008772 Train-accuracy=0.6034375
> 2017-05-26 13:46:49,059 Node[0] Epoch[1] Batch[150] Train-lr=0.0001 Train-iter_per_sec=1.07 Train-samples_per_sec=68.77 Train-loss=-0.0001785517042 Train-accuracy=0.6221875
> 2017-05-26 13:47:35,596 Node[0] Epoch[1] Batch[200] Train-lr=0.0001 Train-iter_per_sec=1.07 Train-samples_per_sec=68.76 Train-loss=0.0003049087618 Train-accuracy=0.5928125
> 2017-05-26 13:48:08,183 Node[0] Epoch[1] Train-loss=-0.000430
> 2017-05-26 13:48:08,184 Node[0] Epoch[1] Train-accuracy=0.633929
> 2017-05-26 13:48:08,185 Node[0] Epoch[1] Time-cost=219.782
> 2017-05-26 13:48:10,143 Node[0] Epoch[1] Validation-loss=0.0001243009465 Validation-accuracy=0.7421875
> 2017-05-26 13:48:57,773 Node[0] Epoch[2] Batch[50] Train-lr=0.0001 Train-iter_per_sec=1.85 Train-samples_per_sec=68.77 Train-loss=0.0001136356458 Train-accuracy=0.6409313725
> 2017-05-26 13:49:44,326 Node[0] Epoch[2] Batch[100] Train-lr=0.0001 Train-iter_per_sec=1.07 Train-samples_per_sec=68.74 Train-loss=-0.0002889547544 Train-accuracy=0.619375
> 2017-05-26 13:50:30,855 Node[0] Epoch[2] Batch[150] Train-lr=0.0001 Train-iter_per_sec=1.07 Train-samples_per_sec=68.78 Train-loss=-0.0003283168329 Train-accuracy=0.6315625
> 2017-05-26 13:51:17,405 Node[0] Epoch[2] Batch[200] Train-lr=0.0001 Train-iter_per_sec=1.07 Train-samples_per_sec=68.75 Train-loss=0.0001538071374 Train-accuracy=0.6328125
> 2017-05-26 13:51:49,995 Node[0] Epoch[2] Train-loss=-0.000471
> 2017-05-26 13:51:49,996 Node[0] Epoch[2] Train-accuracy=0.613393
> 2017-05-26 13:51:49,997 Node[0] Epoch[2] Time-cost=219.844
> 2017-05-26 13:51:51,938 Node[0] Epoch[2] Validation-loss=-0.01205513696 Validation-accuracy=0.78515625
> 2017-05-26 13:52:39,544 Node[0] Epoch[3] Batch[50] Train-lr=0.0001 Train-iter_per_sec=1.85 Train-samples_per_sec=68.78 Train-loss=-2.118847652e-05 Train-accuracy=0.6357230392
> 2017-05-26 13:53:26,107 Node[0] Epoch[3] Batch[100] Train-lr=0.0001 Train-iter_per_sec=1.07 Train-samples_per_sec=68.73 Train-loss=-0.0001829522732 Train-accuracy=0.6059375

**Comment by jmerkow** [comment]
I have been digging into this a bit. Here is what's going on as far as I can tell, please correct me if I am wrong. 
`'loss'` eval metric calculates the [sum of the forward pass](https://github.com/dmlc/mxnet/blob/master/python/mxnet/metric.py#L922), however, the SVMOutput symbol doesn't actually calculate loss, instead, [it passes its input on](https://github.com/dmlc/mxnet/blob/master/src/operator/svm_output-inl.h#L66). So the output of SVMOutput `Forward` is the output of the preceding layer.

Wouldn't this lead to an error in the backward pass? The loss would be a sum of the previous layer's outputs rather than a true loss, this would lead to a miscalculation in the gradient?
Or are forward-loss never actually used, but rather gradients and loss are calculated simultaneously during `Backward`? relevant code: [[1](https://github.com/dmlc/mxnet/blob/master/src/operator/svm_output-inl.h#L91-L95), [2](https://github.com/dmlc/mxnet/blob/9bd0f8915b5f23e407c98086df05246c5905285d/src/operator/svm_output.cc#L10-L49)]

**Comment by yuruofeifei** [comment]
Hi @jmerkow, 

As long as the backward derivation is correct, we don't need loss in forward pass.

But the current way that mxnet use to calculate loss are lack of some info (like margin defined in SVM layer), which is inconvenient, I think it would be better to calculate loss in forward pass as well.

I have created a pr for hinge loss #6496, negative loss does not make sense, you can use by setting eval_metric = [mx.metric.SVMLoss(), ...]

**Comment by jmerkow** [comment]
Thank you @yuruofeifei. 

So I'm understanding correctly, the forward out of the terminal layer is not used at all? I.e. If I set the output to be random numbers (one for each item in the batch), as long as the backward is correct it does not matter?  
The convention with MXNet is to calculate the loss value simultaneously with gradient in backward?

**Comment by yuruofeifei** [comment]
Hi @jmerkow 
SVM is a linear classifier with form WX + b and +1, -1 classification.
From the tutorial, there is a fully connect layer before svm layer, in current mxnet implementation, the output just forward the previous fully connect layer output(in the code F Identity). Maybe in other implementation or framework, the svm layer may store W, b and do the affine transform itself. 

From what I observe(I only looked at svmoutput), MXNet does not compute loss explicitly in the executor itself, so that's why we add the metric utils.

But I think it's not still convenient to compute loss due to lack of info in that way, it's easier to compute loss during forward pass.

**Comment by jmerkow** [comment]
I agree, it should be simple to update the forward pass of SVMOutput symbol. I'll look into it next week.

**Comment by hdriqi** [comment]
Hello I use the hinge loss as metric, but my accuracy still sucks

My architecture is CNN + SVM as output layer. The dataset is cifar-10. The max accuracy i got is 0.1
When I use softmax as output layer, I can achieve 80+ accuracy.

Any suggestion?

**Comment by szha** [comment]
This issue is closed due to lack of activity in the last 90 days. Feel free to ping me to reopen if this is still an active issue. Thanks!
